test/server/execution_event_bus.spec.ts

```diff
@@ -0,0 +1,380 @@
+
+import {
+  DefaultExecutionEventBus,
+  AgentExecutionEvent,
+} from '../../src/server/events/execution_event_bus.js';
+import { Message } from '../../src/types.js';
+
+describe('DefaultExecutionEventBus', () => {
+  let eventBus: DefaultExecutionEventBus;
+
+  beforeEach(() => {
+    eventBus = new DefaultExecutionEventBus();
+  });
+
+  const createMessage = (() => {
+    let counter = 0;
+    return (text: string): Message => ({
+      kind: 'message',
+      messageId: `msg-${counter++}`,
+      role: 'agent',
+      parts: [{ kind: 'text', text }],
+    });
+  })();
+
+  describe('publish and event listeners', () => {
+    it('should emit events to listeners registered with on()', () => {
+      const receivedEvents: AgentExecutionEvent[] = [];
+      const message = createMessage('test');
+
+      eventBus.on('event', (event) => {
+        receivedEvents.push(event);
+      });
+
+      eventBus.publish(message);
+
+      expect(receivedEvents).to.have.length(1);
+      expect(receivedEvents[0]).to.deep.equal(message);
+    });
+
+    it('should emit events to multiple listeners in registration order', () => {
+      const order: string[] = [];
+
+      eventBus.on('event', () => order.push('first'));
+      eventBus.on('event', () => order.push('second'));
+      eventBus.on('event', () => order.push('third'));
+
+      eventBus.publish(createMessage('test'));
+
+      expect(order).to.deep.equal(['first', 'second', 'third']);
+    });
+
+    it('should support registering the same listener multiple times', () => {
+      let callCount = 0;
+      const listener = () => {
+        callCount++;
+      };
+
+      eventBus.on('event', listener);
+      eventBus.on('event', listener);
+      eventBus.on('event', listener);
+
+      eventBus.publish(createMessage('test'));
+
+      expect(callCount).to.equal(3);
+    });
+  });
+
+  describe('finished event', () => {
+    it('should emit finished event to listeners', () => {
+      let finishedCalled = false;
+
+      eventBus.on('finished', () => {
+        finishedCalled = true;
+      });
+
+      eventBus.finished();
+
+      expect(finishedCalled).to.be.true;
+    });
+
+    it('should support multiple finished listeners', () => {
+      const order: string[] = [];
+
+      eventBus.on('finished', () => order.push('first'));
+      eventBus.on('finished', () => order.push('second'));
+
+      eventBus.finished();
+
+      expect(order).to.deep.equal(['first', 'second']);
+    });
+  });
+
+  describe('off() - listener removal', () => {
+    it('should remove a listener', () => {
+      let callCount = 0;
+      const listener = () => {
+        callCount++;
+      };
+
+      eventBus.on('event', listener);
+      eventBus.off('event', listener);
+
+      eventBus.publish(createMessage('test'));
+
+      expect(callCount).to.equal(0);
+    });
+
+    it('should remove listeners in LIFO order when same listener registered multiple times', () => {
```

Comment:
This test still passes if `pop()` is replaced with `shift()` in the code, so it doesn't ensure LIFO order.

However this LIFO behavior isn't mentioned in [Node docs for `EventEmitter`](https://nodejs.org/docs/latest-v18.x/api/events.html#emitterremovelistenereventname-listener):

> `removeListener()` will remove, at most, one instance of a listener from the listener array

I believe we should remove mentions of LIFO order, also I doubt we can distinguish it from the outside for same `listener` instance.